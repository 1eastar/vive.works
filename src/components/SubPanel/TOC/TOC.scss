.TOCItemWrapper {
  display: flex;
  flex-direction: column;
}