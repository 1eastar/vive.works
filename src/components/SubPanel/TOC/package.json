{
  "name": "TOC",
  "version": "0.0.0",
  "private": true,
  "main": "TOC.tsx"
}