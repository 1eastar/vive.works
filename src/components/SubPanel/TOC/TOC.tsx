/* External */
import { useCallback, useMemo, useState } from 'react'
import { useStaticQuery, graphql } from 'gatsby'

/* Internal */
import { isBrowser } from '@utils/browser.utils'
import { getTOCfromMDX, findCurrentMDX } from '@utils/TOC.utils'
import { TOCQueryResult } from '@commons/types/QueryType'
import useIntersectionObserver from '@hooks/useIntersectionObserver.hook'
import TOCItem from './TOCItem'
import * as styles from './TOC.scss'

function TOC() {
  const pathname = isBrowser ? window.location.pathname : ''
  
  const [currentHeading, setCurrentHeading] = useState<Element>()

  const queryData = useStaticQuery<TOCQueryResult>(
    graphql`
      query {
        allMdx {
          nodes {
            body
            frontmatter {
              slug
            }
          }
        }
      }
    `
  )

  const setIntersectTitle = useCallback((target: Element) => setCurrentHeading(target), [])

  useIntersectionObserver(setIntersectTitle, { rootMargin: '0px 0px -60%' })

  const TOCItemList = useMemo(() => {
    const currentMDX = findCurrentMDX(queryData, pathname)
    return currentMDX ? getTOCfromMDX(currentMDX) : []
  }, [queryData, pathname])

  return (
    <div
      className={styles.TOCItemWrapper}
      data-testid='TOC-container'
    >
      { TOCItemList.map(item => (
        <TOCItem
          key={item.title}
          item={item}
          currentHeading={currentHeading}
        />
      )) }
    </div>
  )
}

export default TOC
