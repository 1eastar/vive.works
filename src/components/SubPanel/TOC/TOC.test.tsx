/* External */
import { render, cleanup } from '@testing-library/react'

/* Internal */
import TOC from './TOC'

beforeAll(() => {
  const mockIntersectionObserver = jest.fn().mockReturnValue({
    observe: () => null,
    unobserve: () => null,
    disconnect: () => null
  })

  window.IntersectionObserver = mockIntersectionObserver
  window.location.pathname = '/mockSlug3'
})

afterEach(cleanup)

describe('<TOC/>', () => {
  test('성공적으로 렌더링된다.', () => {
    const { TOCEl } = renderTOC()

    expect(TOCEl()).toBeInTheDocument()
  })

  test('[snapshot] title(h1, h2, h3)이 있을 때 TOC 리스트를 제대로 렌더링한다.', () => {
    const { TOCEl } = renderTOC()

    expect(TOCEl()).toMatchSnapshot()
  })
})


/* Helper function */
function renderTOC() {
  const result = render(<TOC/>)

  const TOCEl = () => result.getByTestId('TOC-container')

  return {
    result,
    TOCEl,
  }
}
