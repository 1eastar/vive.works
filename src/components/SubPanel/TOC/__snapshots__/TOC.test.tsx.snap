// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TOC/> [snapshot] title(h1, h2, h3)이 있을 때 TOC 리스트를 제대로 렌더링한다. 1`] = `
<div
  data-testid="TOC-container"
>
  <a
    class=""
    data-testid="TOCItem-first title"
    href="#first title"
  >
     
    first title
  </a>
  <a
    class=""
    data-testid="TOCItem-second title"
    href="#second title"
  >
      
     
    second title
  </a>
  <a
    class=""
    data-testid="TOCItem-third title"
    href="#third title"
  >
        
     
    third title
  </a>
  <a
    class=""
    data-testid="TOCItem-fourth title"
    href="#fourth title"
  >
      
     
    fourth title
  </a>
  <a
    class=""
    data-testid="TOCItem-fifth title"
    href="#fifth title"
  >
     
    fifth title
  </a>
</div>
`;
